trulens_eval/trulens_eval/feedback/provider/endpoint/dummy.py

```diff
@@ -0,0 +1,388 @@
+from pprint import PrettyPrinter
+import random
+import sys
+from time import sleep
+from typing import Any, Callable, Dict, Optional, Sequence, Tuple, TypeVar
+
+from numpy import random as np_random
+import pydantic
+from pydantic import Field
+
+from trulens_eval.feedback.provider.endpoint.base import DEFAULT_RPM
+from trulens_eval.feedback.provider.endpoint.base import Endpoint
+from trulens_eval.feedback.provider.endpoint.base import EndpointCallback
+from trulens_eval.feedback.provider.endpoint.base import INSTRUMENT
+from trulens_eval.utils.python import locals_except
+from trulens_eval.utils.python import safe_hasattr
+from trulens_eval.utils.serial import JSON
+from trulens_eval.utils.threading import DEFAULT_NETWORK_TIMEOUT
+
+logger = logging.getLogger(__name__)
+
+pp = PrettyPrinter()
+
+A = TypeVar("A")
+B = TypeVar("B")
+T = TypeVar("T")
+
+
+class DummyAPI(pydantic.BaseModel):
+    """A dummy model evaluation API used by DummyEndpoint.
+
+    This is meant to stand in for classes such as OpenAI.completion . Methods in
+    this class are instrumented for cost tracking testing.
+    """
+
+    loading_prob: float
+    """How often to produce the "model loading" response that huggingface api
+    sometimes produces."""
+
+    loading_time: Callable[[DummyAPI],
+                           float] = lambda s: s.random.uniform(0.73, 3.7)
+    """How much time to indicate as needed to load the model."""
+
+    error_prob: float
+    """How often to produce an error response."""
+
+    freeze_prob: float
+    """How often to freeze instead of producing a response."""
+
+    overloaded_prob: float
+    """How often to produce the overloaded message that huggingface sometimes produces."""
+
+    alloc: int
+    """How much data in bytes to allocate when making requests."""
+
+    delay: float = 0.0
+    """How long to delay each request."""
+
+    seed: int = 0xdeadbeef
+    """Control randomness."""
+
+    random: Any = Field(exclude=True)
+    """Random number generator."""
+
+    np_random: Any = Field(exclude=True)
+    """Numpy Random number generator."""
+
+    def __init__(
+        self,
+        error_prob: float = 0 / 100,
+        freeze_prob: float = 0 / 100,
+        overloaded_prob: float = 1 / 100,
+        loading_prob: float = 1 / 100,
+        alloc: int = 1024 * 1024,
+        delay: float = 0.0,
+        seed: int = 0xdeadbeef,
+        **kwargs
+    ):
+        assert error_prob + freeze_prob + overloaded_prob + loading_prob <= 1.0, "Probabilites should not exceed 1.0 ."
+        assert alloc >= 0
+        assert delay >= 0.0
+
+        super().__init__(
+            **locals_except("self", "kwargs"),
+            random=random.Random(seed),
+            np_random=np_random.RandomState(seed)
+        )
+
+    def post(
+        self, url: str, payload: JSON, timeout: Optional[float] = None
+    ) -> Any:
+        """Pretend to make an http post request to some model execution API."""
+
+        if timeout is None:
```

Comment:
Want to allow someone to change DEFAULT_NETWORK_TIMEOUT. If we had it in signature, it will be fixed at the time the definition of post gets loaded hence a subsequent change would not be taken up.